### DISCUSSION TITLE
MQTT Keep Alive(keepalive_backoff) is not disconnecting within 1.5X configured backoff

### DISCUSSION THREAD

**Discussion by ganesh3893** [emqx/emqx#14308] [Q&A]
Hi All,

We are using the EMQX 4.4.18  with 100k active connections.

We have most devices connected with higher keep-alive e.g 15mins whenever any device disconnected due to ungraceful disconnects. Most of the time, EMQX can detect the disconnects but the timing varies and does not match the configured keep-alive configuration. As per the config we have, EMQX should be able to detect the disconnects within 1.5 x of keep-alive but it varies from 1.5x to 3X of keep-alive. we need to provide an exact time to detect the keep-alive disconnects.

Can anyone help us know how we can have consistent behavior and what is defined range time is when detecting ungraceful disconnects?  attached trace logs for the reference with 1 mi
[trace_logs.txt](https://github.com/user-attachments/files/17967912/trace_logs.txt)

bash-5.1$ cat emqx.conf | grep -i keep
## Socket idle keepalive.
rpc.socket_keepalive_idle = 900s
## TCP Keepalive probes interval.
rpc.socket_keepalive_interval = 75s
rpc.socket_keepalive_count = 9
## until 'Keepalive * backoff * 2' timeout.
zone.external.[trace_logs.txt](https://github.com/user-attachments/files/17967900/trace_logs.txt) = 0.75

I tried to override the TCP keep-alive as well but it did not help 
bash-5.1$ env | grep -i keep
EMQX_RPC__SOCKET_KEEPALIVE_IDLE=60s
EMQX_RPC__SOCKET_KEEPALIVE_INTERVAL=15s
EMQX_RPC__SOCKET_KEEPALIVE_COUNT=3
bash-5.1$

**Comment by zmstone** [comment]
Hi @ganesh3893 
please try to change the bacoff from 0.75 to 0.5.

**Reply by terry-xiaoyu** [reply]
```
zone.external.keepalive_backoff = 0.5
```

**Reply by ganesh3893** [reply]
Thanks @zmstone & @terry-xiaoyu 

We change it to 0.5 and it is disconnecting within 1.5x of keep alive.

Do we know what caused this? As per the definition of field zone.external.keepalive_backoff 
Keepalive * backoff * 2 = for 0.5 backoff value, it should have 1x time only right?

https://www.emqx.com/en/blog/mqtt-keep-alive

After introducing the backoff factor, EMQX calculates the maximum timeout using the following formula:

Keepalive * backoff * 2
The default value of backoff is 0.75.

**Reply by zmstone** [reply]
Here is my lengthy clarification

## Background

The `1.5x` interval in the specification was intended to account for "network delays".

However, from a programmer’s perspective, I personally interpreted it as a way to compensate for timing inaccuracies. 
That is, the mechanism allows checks every 0.5x interval, and if no packet is received ~by the third check~ for 2 checks,
then the client is disconnected.
This means the disconnect may occur between `1 x KA-interval` and `1.5 x KA-interval`.

That said, the specification is explicit: the disconnect interval should be exactly `1.5x` the KA interval, leaving no ambiguity.

## Intended Design in EMQX

Ideally EMQX wants to implement aliveness checks using a timer (for each client) which ticks at a user-defined interval (`N` seconds). 
EMQX also wants to allow users to override the `1.5x` interval specified in the protocol to set a custom disconnect time. 
Let’s call this time `D` (default value of which is calculated as 1.5 x KA-interval).
As a result, the disconnect occurs after `D` but before `D+N` seconds of inactivity.

For example:

- If the check interval is set to `10s`, and the desired disconnect time is `30s`, the actual disconnect happens between `30s` and `40s`.
- Using a smaller check interval (e.g., `1s`) improves accuracy, ensuring disconnects occur closer to `30s`.

## The Issue in v4 (Fixed in v5)

In version 4, EMQX used a single factor (default = `0.75`) to calculate both:

1. The check interval (`N`) as `KA-interval * 0.75`.
2. The expiration time (`D`) as `KA-interval * 0.75 * 2`.

This led to disconnects occurring between `1.5 x KA-interval` and `2.25 x KA-interval`.

For example, if the factor is adjusted to `0.5`, disconnects occur between `1 x KA-interval` and `1.5 x KA-interval`.

In your case, my guess is the client have gone idle immediately after the timer started, which could make it seem like the disconnect happened exactly at `1.5x KA-interval` (but in fact slightly earlier, which you may not have noticed).

We currently have no plans to split it into two configs in v4. 
There were discussions about changing the default factor to `0.5`, but no final decision has been made.

## Recommendation

- If the timing of the last packet before idling is uniformly distributed across intervals, theoretically, a factor of `0.625` would fit the best if the goal is to disconnect as close as possible to `1.5x` KA-interval.

- To ensure disconnects occur "exactly" at `1.5 x KA-interval` as per the specification, you’ll need to upgrade to version 5.
In v5, you can configure the multiplier to `1.5` and set the check interval to the smallest allowed unit (`1s`) for maximum accuracy.
However, keep in mind that making timer to tick more frequently means more CPU resource needed.
